### PR TITLE
Mount service source code into corresponding containers

### PR DESCRIPTION
**Pull request by rlucioni**
Also ties in the clone script: repos are cloned if they don't exist when starting and opening devstack.

ECOM-6583

@clintonb @davec-edx @jmbowman

### CODE REVIEW

**Review by jmbowman — APPROVED** [review]

File: Makefile [hunk 1]
```diff
@@ -1,25 +1,28 @@
...
 	@echo "Please use \`make <target>' where <target> is one of"
 	@perl -nle'print $& if m{^[\.a-zA-Z_-]+:.*?## .*$$}' $(MAKEFILE_LIST) | sort | awk 'BEGIN {FS = ":.*?## "}; {printf "\033[36m  %-25s\033[0m %s\n", $$1, $$2}'
 
+clone: ## Clone service repos
+	./clone.py
+
 # TODO Print out help for this target. Even better if we can iterate over the services in docker-compose.yml, and
 # print the actual service names.
-devstack.open.%:    ## Open a shell into the specified service container
+devstack.open.%: clone   ## Open a shell into the specified service container
```
Comment (clintonb) [inline Makefile:12]: This target assumes you have services already running. I don't think it makes sense to clone at this point.

File: clone.py [hunk 2]
```diff
@@ -52,27 +53,29 @@ def __init__(self, clone_url):
...
 
+        parent_path = dirname(dirname(abspath(__file__)))
+        self.path = join(parent_path, self.name)
+
     def clone(self):
         """Clone the repo."""
-        parent_path = dirname(dirname(abspath(__file__)))
-        clone_path = join(parent_path, self.name)
-        subprocess.run(['git', 'clone', self.clone_url, clone_path], check=True)
+        subprocess.run(['git', 'clone', self.clone_url, self.path], check=True)
```
Comment (davec-edx) [inline clone.py:61]: Is it worth sending stdout/stderr to a log file?  or opening a pipe and trap calledProcessError then log or somehow report it?  If it fails, I wouldn't want to lose the cause.

File: Makefile (hunk 1 above)
Comment (rlucioni) [reply Makefile:12]: @clintonb so would you only do the cloning check for `devstack.start`?

File: clone.py (hunk 2 above)
Comment (rlucioni) [reply clone.py:61]: @davec-edx reporting on failures is what the `check=True` is for. See https://docs.python.org/3/library/subprocess.html#subprocess.run. https://github.com/edx/devstack/issues/10 covers figuring out what to do about log aggregation.

File: Makefile (hunk 1 above)
Comment (clintonb) [reply Makefile:12]: Yep.

File: clone.py (hunk 2 above)
Comment (davec-edx) [reply clone.py:61]: Ok, I had read the docs on it, but haven't used the call, I believe it's new for 3.5.  However there is this line:
    ...and stdout and stderr if they were captured.

File: Makefile (hunk 1 above)
Comment (rlucioni) [reply Makefile:12]: Addressed in https://github.com/edx/devstack/pull/16.

File: clone.py (hunk 2 above)
Comment (rlucioni) [reply clone.py:61]: Yes, `run` is new in 3.5. If something goes wrong, we'll see the messages printed to stderr.

Comment (davec-edx) [reply clone.py:61]: 👍

**Pull request merged**